File: ruby/food-chain/food_chain.rb
class FoodChain
  Animal = Struct.new(:name, :tag, :that, :apexPredator?, keyword_init: true)

  CHAIN = [
    nil,
    Animal.new(name: 'fly'),
    Animal.new(name: 'spider',
               that: ' that wriggled and jiggled and tickled inside her',
               tag: "It wriggled and jiggled and tickled inside her.\n"),
    Animal.new(name: 'bird',   tag: "How absurd to swallow a bird!\n"),
    Animal.new(name: 'cat',    tag: "Imagine that, to swallow a cat!\n"),
    Animal.new(name: 'dog',    tag: "What a hog, to swallow a dog!\n"),
    Animal.new(name: 'goat',   tag: "Just opened her throat and swallowed a goat!\n"),
    Animal.new(name: 'cow',    tag: "I don't know how she swallowed a cow!\n"),
    Animal.new(name: 'horse',  tag: "She's dead, of course!\n", apexPredator?: true)
  ].freeze

  def self.song
    new.song
  end

  def initialize
    @song = CHAIN.length.times.map { |i| verse(i) }.join("\n").lstrip
  end
  attr_reader :song

  private

  def verse(n)
    return if n.zero?
    animal = CHAIN[n]
    verse = i_know(animal)
    unless animal.apexPredator?
      n.downto(2) { |i| verse << hunt(CHAIN[i], CHAIN[i - 1]) }
      verse << i_dont_know
    end
    verse
  end

  def i_know(animal)
    "I know an old lady who swallowed a #{animal.name}.\n#{animal.tag || ''}"
  end

  def i_dont_know
    "I don't know why she swallowed the fly. Perhaps she'll die.\n"
  end

  def hunt(predator, prey)
    "She swallowed the #{predator.name} to catch the #{prey.name}#{prey.that || ''}.\n"
  end
end

File: go/food-chain/food_chain.go
package foodchain

import "strings"

var (
	critters      = []string{"fly", "spider", "bird", "cat", "dog", "goat", "cow", "horse"}
	critterAction = map[string]string{
		"spider": "It wriggled and jiggled and tickled inside her.",
		"bird":   "How absurd to swallow a bird!",
		"cat":    "Imagine that, to swallow a cat!",
		"dog":    "What a hog, to swallow a dog!",
		"goat":   "Just opened her throat and swallowed a goat!",
		"cow":    "I don't know how she swallowed a cow!",
		"horse":  "She's dead, of course!",
	}
	critterExtra = map[string]string{
		"spider": " that wriggled and jiggled and tickled inside her",
	}
	whoDunnit = "horse" // spoiler alert!
)

func Verse(v int) string {
	var (
		critter = critters[v-1]
		verse   = strings.Builder{}
	)

	verse.WriteString("I know an old lady who swallowed a ")
	verse.WriteString(critter)
	verse.WriteRune('.')

	if action, ok := critterAction[critter]; ok {
		verse.WriteRune('\n')
		verse.WriteString(action)
	}

	if critter != whoDunnit {
		buildChain(v, &verse)
		verse.WriteRune('\n')
		verse.WriteString("I don't know why she swallowed the fly. Perhaps she'll die.")
	}

	return verse.String()
}

func buildChain(v int, verse *strings.Builder) {
	for i := v - 1; i > 0; i-- {
		verse.WriteRune('\n')
		verse.WriteString("She swallowed the ")
		verse.WriteString(critters[i])
		verse.WriteString(" to catch the ")
		verse.WriteString(critters[i-1])
		verse.WriteString(critterExtra[critters[i-1]])
		verse.WriteRune('.')
	}
}

func Verses(start, end int) string {
	verses := make([]string, 0, end-start+1)
	for v := start; v <= end; v++ {
		verses = append(verses, Verse(v))
	}
	return strings.Join(verses, "\n\n")
}

func Song() string {
	return Verses(1, len(critters))
}

// bench
//
// I'm finding that not using `fmt.Sprintf` and not concatenating strings with `+`
// is reducing the allocs/op and B/op values
//
// BenchmarkSong 	  248278	      4910 ns/op	    8464 B/op	      32 allocs/op
